oak_functions/loader/src/server.rs

```diff
@@ -335,6 +443,95 @@ fn check_export_function_signature(
+/// function.
+///
+/// If the policy is `None`, this function just runs the given function and returns its result.
+/// If a policy is present, the given function is run, but its generated response is only returned
+/// if allowed by the policy. If the execution of the function takes longer than allowed by the
+/// policy, or the resulting response is larger than allowed by the policy, an error response
+/// generated by the policy is returned instead of the actual response.
+async fn apply_policy<F, S>(policy: Option<Policy>, function: F) -> anyhow::Result<Response<Body>>
+where
+    F: std::marker::Send + 'static + FnOnce() -> S,
+    S: std::future::Future<Output = anyhow::Result<FunctionsResponse>> + std::marker::Send,
+{
+    match policy {
+        None => Response::<Body>::try_from(function().await?).context("couldn't get the response"),
+        Some(policy) => {
+            // Use tokio::spawn to actually run the tasks in parallel, for more accurate measurement
+            // of time.
+            let task = tokio::spawn(async move { function().await });
+            tokio::time::sleep(policy.constant_processing_time).await;
+
+            let response = task.now_or_never();
+
+            match response {
+                None => policy.create_error_response("response not available"),
+                Some(response) => match response {
+                    Err(err) => policy.create_error_response(format!("{}", err).as_str()),
+                    Ok(response) => match response {
+                        Err(err) => policy.create_error_response(format!("{}", err).as_str()),
+                        Ok(rsp) => {
+                            // All responses from Oak-Functions have an additional
+                            // `oak-policy-padding` header.
+                            let size = estimate_size(&rsp)?;
+                            match size.cmp(&(policy.constant_response_size_bytes as i64)) {
+                                Ordering::Greater => {
+                                    policy.create_error_response("response too large")
+                                }
+                                Ordering::Less => policy.pad_response(rsp),
+                                Ordering::Equal => Response::<Body>::try_from(rsp)
+                                    .context("couldn't get the response"),
+                            }
+                        }
+                    },
+                },
+            }
+        }
+    }
+}
+
+/// Estimates the size of the given response.
+///
+/// Measures the size of the response when formatted according to
+/// [RFC-2616](https://tools.ietf.org/html/rfc2616#section-6).
+/// The only relevant parts of the response are the status code and the response body. We also add
+/// the `oak-policy-padding` header to all responses. This header name is therefore also included
+/// when estimating the size.
+fn estimate_size(response: &FunctionsResponse) -> anyhow::Result<i64> {
```

Comment:
I'm not sure why the size of the HTTP response needs to be estimated at all, and also why it is estimated in this particular way, and what does it mean for it to be an estimate; also the logic seems quite brittle, e.g. it would stop working for HTTP2, right?

I expect the logic to be simply that the sum of the body and the padding header is constant; no other details should be necessary, in particular no estimation is needed. Or am I missing some case in which that does not work?

Reply:
Different status codes have different lengths. If we apply the size only to the body and the padding header, then the total size of the transmitted response could be different for different status codes. I think in general, for the same request we can still respond with 500 (INTERNAL_SERVER_ERROR) in some cases instead of 200. So, I think the status code should be included in the calculation of the size. 

Of course, if we want to support HTTP/2 this functionality has to be extended. But it is implementing the standard. I would not call it brittle.

It is estimated, because the final response has additional headers like `content-length`, which are not included in the calculation of the size here. I think that is what hyper adds to the response. And I think the size of those additional headers is constant.

Reply:
Ah thanks, I didn't realise that the status code description actually gets sent back too! I thought it was just the numeric code.

I wonder if we should reconsider the overall approach slightly; instead of basing it on HTTP semantics, we may want our own Request and Response objects, and always return an HTTP OK response.

For instance, the Request message may encapsulate the request body and the policy, and the Response message may encapsulate the response body, the response code, the padding, and eventually additional information (e.g. logs to be sent to the client). Let's not block on it for this PR, but WDYT about that?

Reply:
I did think of that, but then we have to publish the schema for the request and the response body, and the client has to be able to deserialize the body, to be able to get the actual response and fail/success status. But I guess that should be fine. Created #2032.

Reply:
We are effectively using an implicit schema by referring to header names and status codes anyways, so we may as well have an explicit one instead (e.g. either protobuf or JSON). Thanks for creating the issue, let's follow up there.